Reword the read_from function and add documentation to memory_model

xbauch opened the pull request:
Also separate the inner-most loop into dedicated function that handles a single
r-w events pair.

No functional change was introduced.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig commented:
(I am aware this is still in draft state)  Being at fault for this messy code, I really appreciate this cleanup and improvement!! The only thing I'd ask for are breaking up changes into a few small commits (in this single PR) to make reviewing easier.

Commit:
Use ranged for-loops

Shorter and maybe easier to read.

Commit:
Separate innermost-loop of read_from

To a dedicated function `register_read_from_choice_symbol` that returns the
choice symbol to be used in the big disjunction.

Commit:
Wrap generalised disjunction into a separate function

And use in building the `rf_some` expression.

Commit:
Rename variables

To remove redundancy and improve readability.

Commit:
Remove continue-jump out of loops

By negating the jump condition.

Commit:
Inline constraints

Their form is simple enough to justify having them nameless.

Commit:
Add documentation to memory_model_baset functions

tautschnig commented on src/goto-symex/memory_model.cpp:
```diff
@@ -49,30 +49,21 @@ void memory_model_baset::read_from(symex_target_equationt &equation)
+  for(const auto &address : address_map)
   {
-    const a_rect &a_rec=a_it->second;
+    const a_rect &a_rec=address.second;
```
clang-format will want space around `=` here.

xbauch replied:
Done.

tautschnig commented on src/goto-symex/memory_model.cpp:
```diff
@@ -120,3 +94,38 @@ void memory_model_baset::read_from(symex_target_equationt &equation)
+  symex_target_equationt &equation)
+{
+  bool is_rfi=
+    w->source.thread_nr==r->source.thread_nr;
```
While you are at it: I'd move this further down, just right before the first use of `is_rfi`.

xbauch replied:
Done

tautschnig commented on src/goto-symex/memory_model.cpp:
```diff
@@ -74,18 +74,12 @@ void memory_model_baset::read_from(symex_target_equationt &equation)
 
       // value equals the one of some write
-      exprt rf_some;
+      exprt rf_some = big_or(rf_some_operands);
```
1) There is `disjunction` in `std_expr.h`, which you could use instead of introducing `big_or`.
2) I'd now move this assignment and the comment below the next block as `rf_some` will not be needed if the `continue` is hit.

xbauch replied:
I did (1) but wrt. (2) there is not `continue` in the latest version. Is it ok?

tautschnig replied:
Indeed, but you could still move it into the body of the if statement?

xbauch replied:
You mean in that particular commit? Because in the latest version, it's already there.

tautschnig replied:
Ah, my apologies, I had missed that. Perfectly fine as-is now.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 56e73690112b524203a563e935d50fe5e4edce5c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104538802
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

Commit:
Fix clang format

No functional change.
